File: data/translated_projects/commons-fileupload/src/main/org/apache/commons/fileupload/ParameterParser.py
from __future__ import annotations
import io
import typing
from typing import *
import os
from src.main.org.apache.commons.fileupload.util.mime.MimeUtility import *
import java # type: ignore
from src.main.org.apache.commons.fileupload.java_handler import *
class ParameterParser(metaclass=StaticFieldRedirector):
    def parse3(self, charArray: typing.List[str], offset: int, length: int, separator: str) -> typing.Dict[str, str]:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = charArray
        translatedArg1 = JavaHandler.valueToObject(offset, "int", idMapPyToJ)
        translatedArg2 = JavaHandler.valueToObject(length, "int", idMapPyToJ)
        translatedArg3 = JavaHandler.valueToObject(separator, "char", idMapPyToJ)
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parse3(translatedArg0, translatedArg1, translatedArg2, translatedArg3), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, charArray)
            JavaHandler.mapping(translatedArg1, idMapJToPy, offset)
            JavaHandler.mapping(translatedArg2, idMapJToPy, length)
            JavaHandler.mapping(translatedArg3, idMapJToPy, separator)
    def parse2(self, charArray: typing.List[str], separator: str) -> typing.Dict[str, str]:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = charArray
        translatedArg1 = JavaHandler.valueToObject(separator, "char", idMapPyToJ)
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parse2(translatedArg0, translatedArg1), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, charArray)
            JavaHandler.mapping(translatedArg1, idMapJToPy, separator)
    def parse1(self, str_: str, separator: str) -> typing.Dict[str, str]:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = JavaHandler.valueToObject(str_, "String", idMapPyToJ)
        translatedArg1 = JavaHandler.valueToObject(separator, "char", idMapPyToJ)
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parse1(translatedArg0, translatedArg1), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, str_)
            JavaHandler.mapping(translatedArg1, idMapJToPy, separator)
    def parse0(self, str_: str, separators: typing.List[str]) -> typing.Dict[str, str]:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = JavaHandler.valueToObject(str_, "String", idMapPyToJ)
        translatedArg1 = separators
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parse0(translatedArg0, translatedArg1), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, str_)
            JavaHandler.mapping(translatedArg1, idMapJToPy, separators)
    def setLowerCaseNames(self, b: bool) -> None:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = JavaHandler.valueToObject(b, "boolean", idMapPyToJ)
        idMapJToPy = dict()
        try:
            self.javaObj.setLowerCaseNames(translatedArg0)
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, b)
    def isLowerCaseNames(self) -> bool:

        idMapPyToJ = self.javaObj.pyToJ()
        
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.isLowerCaseNames(), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            
    def __init__(self) -> None:

        idMapPyToJ = JavaHandler.valueToObject(dict(), "Map")
        
        idMapJToPy = dict()
        try:
            self.javaObj = ParameterParser.javaClz()
            self.javaObj.setPythonObject(self)
            self.javaObj.jToPy(idMapJToPy)
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            
    def __parseQuotedToken(self, terminators: typing.List[str]) -> str:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = terminators
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parseQuotedToken(translatedArg0), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, terminators)
    def __parseToken(self, terminators: typing.List[str]) -> str:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = terminators
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.parseToken(translatedArg0), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, terminators)
    def __isOneOf(self, ch: str, charray: typing.List[str]) -> bool:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = JavaHandler.valueToObject(ch, "char", idMapPyToJ)
        translatedArg1 = charray
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.isOneOf(translatedArg0, translatedArg1), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, ch)
            JavaHandler.mapping(translatedArg1, idMapJToPy, charray)
    def __getToken(self, quoted: bool) -> str:

        idMapPyToJ = self.javaObj.pyToJ()
        translatedArg0 = JavaHandler.valueToObject(quoted, "boolean", idMapPyToJ)
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.getToken(translatedArg0), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            JavaHandler.mapping(translatedArg0, idMapJToPy, quoted)
    def __hasChar(self) -> bool:

        idMapPyToJ = self.javaObj.pyToJ()
        
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(self.javaObj.hasChar(), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            self.javaObj.jToPy(idMapJToPy)
            
    javaClz = java.type("org.apache.commons.fileupload.ParameterParser")
    def __getattr__(self, name):
        try:
            return object.__getattribute__(self, name)
        except AttributeError:
            pass
        return JavaHandler.mapping(getattr(self.javaClz, StaticFieldRedirector.unmangle_name(name)))

File: logs/glue/commons-fileupload/src/main/java/org/apache/commons/fileupload/ParameterParser.java
package org.apache.commons.fileupload;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import org.apache.commons.fileupload.ContextInitializer;
import org.apache.commons.fileupload.ExceptionHandler;
import org.apache.commons.fileupload.IntegrationUtils;
import org.apache.commons.fileupload.util.mime.MimeUtility;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

public class ParameterParser {
  private transient Value obj =
      IntegrationUtils.createDefaultPythonObject(
          ContextInitializer.getPythonClass(
              "main/org/apache/commons/fileupload/ParameterParser.py", "ParameterParser"));
  public char[] chars = null;
  public int pos = 0;
  public int len = 0;
  public int i1 = 0;
  public int i2 = 0;
  public boolean lowerCaseNames = false;

  public Value getPythonObject() {
    return obj;
  }

  public void setPythonObject(Value obj) {
    this.obj = obj;
  }

  public Map<String, String> parse3(
      final char[] charArray, int offset, int length, char separator) {

    if (charArray == null) {
      return new HashMap<String, String>();
    }
    HashMap<String, String> params = new HashMap<String, String>();
    this.chars = charArray;
    this.pos = offset;
    this.len = length;

    String paramName = null;
    String paramValue = null;
    while (hasChar()) {
      paramName = parseToken(new char[] {'=', separator});
      paramValue = null;
      if (hasChar() && (charArray[pos] == '=')) {
        pos++; // skip '='
        paramValue = parseQuotedToken(new char[] {separator});

        if (paramValue != null) {
          try {
            paramValue = MimeUtility.decodeText(paramValue);
          } catch (UnsupportedEncodingException e) {
          }
        }
      }
      if (hasChar() && (charArray[pos] == separator)) {
        pos++; // skip separator
      }
      if ((paramName != null) && (paramName.length() > 0)) {
        if (this.lowerCaseNames) {
          paramName = paramName.toLowerCase(Locale.ENGLISH);
        }

        params.put(paramName, paramValue);
      }
    }
    return params;
  }

  public Map<String, String> parse2(final char[] charArray, char separator) {

    if (charArray == null) {
      return new HashMap<String, String>();
    }
    return parse3(charArray, 0, charArray.length, separator);
  }

  public Map<String, String> parse1(final String str, char separator) {

    if (str == null) {
      return new HashMap<String, String>();
    }
    return parse2(str.toCharArray(), separator);
  }

  public Map<String, String> parse0(final String str, char[] separators) {

    if (separators == null || separators.length == 0) {
      return new HashMap<String, String>();
    }
    char separator = separators[0];
    if (str != null) {
      int idx = str.length();
      for (char separator2 : separators) {
        int tmp = str.indexOf(separator2);
        if (tmp != -1 && tmp < idx) {
          idx = tmp;
          separator = separator2;
        }
      }
    }
    return parse1(str, separator);
  }

  public void setLowerCaseNames(boolean b) {

    this.lowerCaseNames = b;
  }

  public boolean isLowerCaseNames() {

    return this.lowerCaseNames;
  }

  public ParameterParser() {

    super();
  }

  public String parseQuotedToken(final char[] terminators) {

    char ch;
    i1 = pos;
    i2 = pos;
    boolean quoted = false;
    boolean charEscaped = false;
    while (hasChar()) {
      ch = chars[pos];
      if (!quoted && isOneOf(ch, terminators)) {
        break;
      }
      if (!charEscaped && ch == '"') {
        quoted = !quoted;
      }
      charEscaped = (!charEscaped && ch == '\\');
      i2++;
      pos++;
    }
    return getToken(true);
  }

  public String parseToken(final char[] terminators) {

    char ch;
    i1 = pos;
    i2 = pos;
    while (hasChar()) {
      ch = chars[pos];
      if (isOneOf(ch, terminators)) {
        break;
      }
      i2++;
      pos++;
    }
    return getToken(false);
  }

  public boolean isOneOf(char ch, final char[] charray) {

    boolean result = false;
    for (char element : charray) {
      if (ch == element) {
        result = true;
        break;
      }
    }
    return result;
  }

  public String getToken(boolean quoted) {

    while ((i1 < i2) && (Character.isWhitespace(chars[i1]))) {
      i1++;
    }
    while ((i2 > i1) && (Character.isWhitespace(chars[i2 - 1]))) {
      i2--;
    }
    if (quoted && ((i2 - i1) >= 2) && (chars[i1] == '"') && (chars[i2 - 1] == '"')) {
      i1++;
      i2--;
    }
    String result = null;
    if (i2 > i1) {
      result = new String(chars, i1, i2 - i1);
    }
    return result;
  }

  public boolean hasChar() {

    return this.pos < this.len;
  }

  public java.util.Map pyToJ() {
    java.util.Map idMap = new java.util.HashMap();
    return pyToJ(idMap);
  }

  public java.util.Map pyToJ(java.util.Map idMap) {

    this.chars =
        (char[])
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__chars"), "char[]", idMap, char.class, chars);
    this.pos =
        (int)
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__pos"), "int", idMap, pos);
    this.len =
        (int)
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__len"), "int", idMap, len);
    this.i1 =
        (int)
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__i1"), "int", idMap, i1);
    this.i2 =
        (int)
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__i2"), "int", idMap, i2);
    this.lowerCaseNames =
        (boolean)
            IntegrationUtils.valueToObject(
                this.obj.getMember("_ParameterParser__lowerCaseNames"),
                "boolean",
                idMap,
                lowerCaseNames);
    return idMap;
  }

  public Value jToPy() {
    Value idMap = IntegrationUtils.mapToPython(new java.util.HashMap());
    return jToPy(idMap);
  }

  public Value jToPy(Value idMap) {

    this.obj.invokeMember("__setattr__", "javaObj", this);
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__chars",
        IntegrationUtils.mapToPython(chars, idMap, this.obj.getMember("_ParameterParser__chars")));
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__pos",
        IntegrationUtils.mapToPython(pos, idMap, this.obj.getMember("_ParameterParser__pos")));
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__len",
        IntegrationUtils.mapToPython(len, idMap, this.obj.getMember("_ParameterParser__len")));
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__i1",
        IntegrationUtils.mapToPython(i1, idMap, this.obj.getMember("_ParameterParser__i1")));
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__i2",
        IntegrationUtils.mapToPython(i2, idMap, this.obj.getMember("_ParameterParser__i2")));
    this.obj.invokeMember(
        "__setattr__",
        "_ParameterParser__lowerCaseNames",
        IntegrationUtils.mapToPython(
            lowerCaseNames, idMap, this.obj.getMember("_ParameterParser__lowerCaseNames")));
    return idMap;
  }
}
